c_string: change from struct to char*

atharvanan1 opened the pull request:
- Changed CString from struct to char* type for easier operatibility
- This lets us treat the CString type as char* type while doing
operations

Commit:
c_string: change from struct to char*

- Changed CString from struct to char* type for easier operatibility
- This lets us treat the CString type as char* type while doing
operations

Commit:
c_string: add errno and const buffer

- add errno when out-of-bound pos is provided
- string_insert_str takes const char buffer

Commit:
c_string: add documentation

roja-p017 commented on src/c_string/c_string.c:
```diff
@@ -2,91 +2,135 @@
-    size_t len = string->size;
+    size_t len = strlen(string);
     if (pos > len) {
+        errno = ENOMEM;
```
Hey I think `EINVAL` would be a better option

atharvanan1 replied:
Done

roja-p017 commented on src/c_string/c_string.c:
```diff
@@ -2,91 +2,135 @@
     }
-    char* new_string_data = (char *)realloc(string->string_data, len);
-    if (new_string_data == NULL) {
+    char* new_string = (char *)realloc(string, len);
```
1. Use `CString` instead of `char*`
2. `len` -> `len+1`

atharvanan1 replied:
I think len will have to be modified to `len + 1` before insert and `len + 2` after insert. Since, strlen will provide us with length excluding NULL character.

atharvanan1 replied:
This is now fixed. Using `alloc_len` as variable

roja-p017 commented on src/c_string/c_string.c:
```diff
@@ -2,91 +2,135 @@
     size_t ins_str_len = strlen(ins_str);
-    char* new_string_data = (char *)realloc(string->string_data, len + ins_str_len);
-    if (new_string_data == NULL) {
+    char* new_string = (char *)realloc(string, len + ins_str_len);
```
Use `CString` instead of `char*`

atharvanan1 replied:
Done!

roja-p017 commented on src/c_string/c_string.c:
```diff
@@ -2,91 +2,135 @@
 
-    string->string_data = new_string_data;
-    string->size = len + ins_str_len;
     char* temp = (char*) malloc(sizeof(char) * len - pos);
```
Use `CString` instead of `char*`

atharvanan1 replied:
Done

Commit:
fix: memory leaks

- Changed the function prototypes to use pointer to CString type
as the address pointed by CString type needs to be updated by the
end of function call. Hence, prototypes now take CString* instead
of CString. This error produced memory leak where the memory address
first assigned using `new` call was persistent through `insert` and
`insert_str` calls. This is wrong, so fixed that.
- Changed `len` variable to `alloc_len` as alloc length is +1 of
length returned by strlen

atharvanan1 commented:
This is implementation complete for now. Will create new PRs for when to add remove, append functions

atharvanan1 commented:
@roja-p017 Let me know what you think about this PR, approve if everything looks good.
